# Adding new features

In the previous chapter we talked about how, from the point of view of TDD-based development, all of the defects can almost be considered as *features* that just weren't defined initially. Another way of looking at it is that they're called *features* if we're asked for them explicitly, and defects if they're implicit in another *feature*, but haven't been developed.

That is, when we say that we want to be able to mark a task as completed, to follow up with out to-do list project, we can assume that the system mustn't break if we try to mark an inexistent task. For this reason, we'd say that this *feature* had a bug, and that's precisely what we fixed in the previous chapter.

But in this chapter, we'll tackle how to add new features to existing software by following a TDD approach. And, as it could be expected, we're not actually going to make any change to our methodology. We'll still be starting with an acceptance test, and delving deeper into the application and the necessary changes.

All in all, it's a different scenario. A new behavior might require us to modify existing software units, and we need to make sure that the changes don't break any of the already created functionality.

## New user story

The next business requirement is to allow to edit an existing task.

### US-4

* As a user
* I want to modify a existing task in the list
* So that, I can express my ideas better

Initially, this story requires the creation of a new *endpoint* with which to change a task's information.

```
PUT /api/todo/{taskId}
```

If our application has a *front-end*, we might need an *endpoint* to recover the information of the task that we wish to edit, with the purpose of filling out the form with the current data. In this case, it would be:

```
GET /api/todo/{taskId}
```

In both cases, the procedure will be the same: we'll begin by creating an acceptance test, initiating the development process. What we will find is that some of the necessary components are already created.

```php
namespace App\Tests\Katas\TodoList;

use Symfony\Bundle\FrameworkBundle\Client;
use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
use Symfony\Component\HttpFoundation\Response;

class TodoListAcceptanceTest extends WebTestCase
{
    private Client $client;

    # ...
    
    public function asUserIWantToModifyAnExistingTask(): void
    {
        $this->givenIHaveAddedTasks(
            [
                'Write a test that fails',
                'Write code to make the test pass',
            ]
        );

        $this->client->request(
            'PUT',
            '/api/todo/2',
            [],
            [],
            ['CONTENT-TYPE' => 'json/application'],
            json_encode(['task' => 'Write production code to make the test pass'], JSON_THROW_ON_ERROR)
        );

        $putResponse = $this->client->getResponse();
        
        self::assertEquals(204, $putResponse->getStatusCode());

        $response = $this->whenIRequestTheListOfTasks();
        $this->thenICanSeeAddedTasksInTheList(
            [
                '[ ] 1. Write a test that fails',
                '[ ] 2. Write production code to make the test pass',
            ],
            $response
        );
    }
    
    # ...
}
```

So, we run the test to see what it tells us. As expected, the *endpoint* cannot be found because we don't have the route, so we start by defining it.

```yaml
api_add_task:
  path: /api/todo
  controller: App\TodoList\Infrastructure\EntryPoint\Api\TodoListController::addTask
  methods: ['POST']

api_get_task_list:
  path: /api/todo
  controller: App\TodoList\Infrastructure\EntryPoint\Api\TodoListController::getTaskList
  methods: ['GET']

api_mark_task_completed:
  path: /api/todo/{taskId}
  controller: App\TodoList\Infrastructure\EntryPoint\Api\TodoListController::markTaskCompleted
  methods: ['PATCH']

api_edit_task:
  path: /api/todo/{taskId}
  controller: App\TodoList\Infrastructure\EntryPoint\Api\TodoListController::modifyTask
  methods: ['PUT']
```

When rerunning the test after this change, it'll tell us that there isn't any action in the controller that's able to respond to this route.

```
"The controller for URI "/api/todo/2" is not callable. Expected method "modifyTask"
```

So we'll have to add a new empty action.

```php
namespace App\TodoList\Infrastructure\EntryPoint\Api;

use App\TodoList\Application\AddTaskHandler;
use App\TodoList\Application\GetTaskListHandler;
use App\TodoList\Application\MarkTaskCompletedHandler;
use App\TodoList\Application\TaskListTransformer;
use InvalidArgumentException;
use OutOfBoundsException;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use function is_string;

class TodoListController
{
    private AddTaskHandler $addTaskHandler;
    private GetTaskListHandler $getTaskListHandler;
    private TaskListTransformer $taskListTransformer;
    private MarkTaskCompletedHandler $markTaskCompletedHandler;

    public function __construct(
        AddTaskHandler $addTaskHandler,
        GetTaskListHandler $getTaskListHandler,
        TaskListTransformer $taskListTransformer,
        MarkTaskCompletedHandler $markTaskCompletedHandler
    ) {
        $this->addTaskHandler = $addTaskHandler;
        $this->getTaskListHandler = $getTaskListHandler;
        $this->taskListTransformer = $taskListTransformer;
        $this->markTaskCompletedHandler = $markTaskCompletedHandler;
    }

    # ...
    
    public function modifyTask(int $taskId, Request $request): Response
    {
        throw new \RuntimeException(sprintf('Implement %s', __METHOD__));
    }

    # ...
}
```

In the new execution of the test, the error will be:

```
RuntimeException: "Implement App\TodoList\Infrastructure\EntryPoint\Api\TodoListController::modifyTask"
```

Which tells us that we have to dive into the unit level to implement this action in the controller. This cycle will ring a bell, because it's what we've been doing in all this part of the book.

But the truth is that this routine is something positive. We always have a concrete task to tackle at every moment, be it creating a test or writing production code, and we don't have to worry about anything else. The acceptance test tells us what to do, and at each level we just have to think about the specific component that we're working on.

It's time for us to implement the controller. As we already know, at this stage we have to design. Basically, it's a similar action to adding a task, but in this case we'll receive the id of the task that we're going to modify, as well as its new description.

We'll need a use case that expresses this user intention, to which we'll pass the two pÃ¬eces of data that we need. If everything goes as planned, we'll return the 204 response (no content).

We add a test that encompasses all this:

```php
namespace App\Tests\TodoList\Infrastructure\EntryPoint\Api;

use App\TodoList\Application\AddTaskHandler;
use App\TodoList\Application\GetTaskListHandler;
use App\TodoList\Application\MarkTaskCompletedHandler;
use App\TodoList\Application\TaskListTransformer;
use App\TodoList\Infrastructure\EntryPoint\Api\TodoListController;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;
use OutOfBoundsException;
use Symfony\Component\HttpFoundation\Request;

class TodoListControllerTest extends TestCase
{
    private const TASK_DESCRIPTION = 'Task Description';
    private const COMPLETED_TASK_ID = 1;
    private const TASK_ID = 1;

    private AddTaskHandler $addTaskHandler;
    private TodoListController $todoListController;
    private GetTaskListHandler $getTaskListHandler;
    private TaskListTransformer $taskListTransformer;
    private MarkTaskCompletedHandler $markTaskCompletedHandler;
    private UpdateTaskHandler $updateTaskHandler;

    protected function setUp(): void
    {
        $this->addTaskHandler = $this->createMock(AddTaskHandler::class);
        $this->getTaskListHandler = $this->createMock(GetTaskListHandler::class);
        $this->taskListTransformer = $this->createMock(TaskListTransformer::class);
        $this->markTaskCompletedHandler = $this->createMock(MarkTaskCompletedHandler::class);
        $this->updateTaskHandler = $this->createMock(UpdateTaskHandler::class);

        $this->todoListController = new TodoListController(
            $this->addTaskHandler,
            $this->getTaskListHandler,
            $this->taskListTransformer,
            $this->markTaskCompletedHandler,
            $this->updateTaskHandler
        );
    }

    # ...
    
    /** @test */
    public function shouldModifyATask(): void
    {
        $this->updateTaskHandler
            ->expects(self::once())
            ->method('execute')
            ->with(self::TASK_ID, self::TASK_DESCRIPTION);

        $request = new Request(
            [],
            [],
            [],
            [],
            [],
            ['CONTENT-TYPE' => 'json/application'],
            json_encode(['task' => self::TASK_DESCRIPTION], JSON_THROW_ON_ERROR)
        );

        $response = $this->todoListController->modifyTask(self::TASK_ID,$request);

        self::assertEquals(204, $response->getStatusCode());
    }

    # ...
}
```

If we run the test, it will ask us to create the `UpdateTaskHandler` use case.

```php
namespace App\TodoList\Application;

class UpdateTaskHandler
{
}
```

And next, it will ask for the `execute` method.

```php
namespace App\TodoList\Application;

class UpdateTaskHandler
{
    public function execute()
    {
        throw new \RuntimeException(sprintf('Implement %s', __METHOD__));
    }
}
```

Once we have that, it again asks us to implement the controller's action. So we get to it:

```php
namespace App\TodoList\Infrastructure\EntryPoint\Api;

use App\TodoList\Application\AddTaskHandler;
use App\TodoList\Application\GetTaskListHandler;
use App\TodoList\Application\MarkTaskCompletedHandler;
use App\TodoList\Application\UpdateTaskHandler;
use App\TodoList\Application\TaskListTransformer;
use InvalidArgumentException;
use OutOfBoundsException;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use function is_string;

class TodoListController
{
    private AddTaskHandler $addTaskHandler;
    private GetTaskListHandler $getTaskListHandler;
    private TaskListTransformer $taskListTransformer;
    private MarkTaskCompletedHandler $markTaskCompletedHandler;
    private UpdateTaskHandler $updateTaskHandler;

    public function __construct(
        AddTaskHandler $addTaskHandler,
        GetTaskListHandler $getTaskListHandler,
        TaskListTransformer $taskListTransformer,
        MarkTaskCompletedHandler $markTaskCompletedHandler,
        UpdateTaskHandler $updateTaskHandler
    ) {
        $this->addTaskHandler = $addTaskHandler;
        $this->getTaskListHandler = $getTaskListHandler;
        $this->taskListTransformer = $taskListTransformer;
        $this->markTaskCompletedHandler = $markTaskCompletedHandler;
        $this->updateTaskHandler = $updateTaskHandler;
    }

    # ...

    public function modifyTask(int $taskId, Request $request): Response
    {
        $payload = $this->obtainPayload($request);

        $this->updateTaskHandler->execute($taskId, $payload['task']);

        return new JsonResponse('', Response::HTTP_NO_CONTENT);
    }

    # ...
}
```

And the controller's unit test passes. If we return to the acceptance test, as we should be doing now, it'll tell us what we have to do next:

```
RuntimeException: "Implement App\TodoList\Application\UpdateTaskHandler::execute"
```

So, it's time to get into the application layer. Again, we have to design this level, which poses an interesting problem.

In principle, we have defined that the field of the task that can be changed is just its description. Therefore, this action has to respect the current stated of the completed flag. So what we want is to recover the stored task, modify its description, and save it.

Therefore, we'll ask the repository for the task, we'll change it, and we'll store it again.

```php
namespace App\Tests\TodoList\Application;

use App\TodoList\Application\UpdateTaskHandler;
use App\TodoList\Domain\Task;
use App\TodoList\Domain\TaskRepository;
use PHPUnit\Framework\TestCase;

class UpdateTaskHandlerTest extends TestCase
{

    private const TASK_ID = 1;

    public function testShouldUpdateATask(): void
    {
        $task = new Task(self::TASK_ID, 'Task Description');
        $taskRepository = $this->createMock(TaskRepository::class);
        $taskRepository
            ->method('retrieve')
            ->with(self::TASK_ID)
            ->willReturn($task);
        $taskRepository
            ->expects(self::once())
            ->method('store')
            ->with(new Task(self::TASK_ID, 'New Task Description'));

        $updateTaskHandler = new UpdateTaskHandler($taskRepository);

        $updateTaskHandler->execute(self::TASK_ID, 'New Task Description');
    }
}
```

When we run the test, it will ask us to implement the use case, as the repository had already been defined previously.

The implementation will surely force us to introduce some new method in `Task`, making a way for the description to be updated. This one, for example:

```php
namespace App\TodoList\Application;

use App\TodoList\Domain\TaskRepository;

class UpdateTaskHandler
{

    private TaskRepository $taskRepository;

    public function __construct(TaskRepository $taskRepository)
    {
        $this->taskRepository = $taskRepository;
    }

    public function execute(int $taskId, string $newTaskDescription): void
    {
        $task = $this->taskRepository->retrieve($taskId);

        $task->updateDescription($newTaskDescription);

        $this->taskRepository->store($task);
    }
}
```

I've chosen this implementation to simplify. However, as I write this, I can come up some ideas that could be interesting in a realistic use case. One of them could be to apply a certain immutability, that is, instead of updating the `Task` object, we'd create a new one filled out with new values.

But we'll leave those refinements for another occasion. If the run the test, it'll tell us that `Task` is lacking the `updateDescription` method, which we'll have to develop with the help of a unit test.

```php
namespace App\Tests\TodoList\Domain;

use App\TodoList\Domain\Task;
use PHPUnit\Framework\TestCase;
use InvalidArgumentException;

class TaskTest extends TestCase
{
    /** @test */
    public function shouldNotAllowEmptyDescription(): void
    {
        $this->expectException(InvalidArgumentException::class);

        new Task(1, '');
    }

    /** @test */
    public function shouldProvideRepresentation(): void
    {
        $expected = '[ ] 1. Task Description';
        $task = new Task(1, 'Task Description');

        $representation = $task->representedAs('[:check] :id. :description');

        self::assertEquals($expected, $representation);
    }

    /** @test */
    public function shouldMarkTaskCompleted(): void
    {
        $expected = '[â] 1. Task Description';
        $task = new Task(1, 'Task Description');
        $task->markCompleted();

        $representation = $task->representedAs('[:check] :id. :description');

        self::assertEquals($expected, $representation);
    }

    /** @test */
    public function shouldUpdateDescription(): void
    {
        $expected = '[ ] 1. New Task Description';
        $task = new Task(1, 'Task Description');
        $task->updateDescription('New Task Description');

        $representation = $task->representedAs('[:check] :id. :description');

        self::assertEquals($expected, $representation);
    }

}
```

To make the test pass, we have to introduce the method.

```php
namespace App\TodoList\Domain;

use InvalidArgumentException;

class Task
{
    private int $id;
    private string $description;
    private bool $completed;

    public function __construct(int $id, string $description)
    {
        $this->id = $id;

        if ($description === '') {
            $exceptionMessage = 'Task description should not be empty';
            throw new InvalidArgumentException($exceptionMessage);
        }

        $this->description = $description;
        $this->completed = false;
    }

    public function id(): int
    {
        return $this->id;
    }

    public function representedAs(string $format): string
    {
        $values = [
            ':check' => $this->completed ? 'â' : ' ',
            ':id' => $this->id,
            ':description' => $this->description
        ];
        return strtr($format, $values);

    }

    public function markCompleted(): void
    {
        $this->completed = true;
    }

    public function updateDescription(string $newTaskDescription): void
    {
        $this->description = $newTaskDescription;
    }
}
```

The test passes, but we've noticed a problem. A few moments ago we had implemented a validation in order to prevent `Task::description` from being an empty string. To ensure that we're fulfilling this business rule, we should introduce another test that verifies it and implement the answer that we want to give to this case.

However, we haven't covered this in either the acceptance or the controller level. What should we do then? Solve it now and add tests at the other levels later, or wait and add this protection in a new iteration?

Personally, I think that the best option is to take note of this and solve it in a new cycle. It's important to focus in the feature that we're developing right now and finish the cycle.

Therefore, when we make the `Task` unit test, we first return to the `UpdateTaskHandler` and verify that it passes, which is exactly what happens.

And having this level in green, we try the acceptance one again, which also passes without any trouble.

The result is that the new story is now implemented, although as we've discovered, we need to do an extra iteration to prevent the problem of trying to update the description of a task with an invalid value.

Could we have prevented this earlier? Well it might be so. However, we'd still have needed to introduce tests at the different levels, just as we did in the previous chapter. The value of using TDD lies, precisely, in developing a series of thought habits and a certain automatization. In other words, to reach a degree of discipline and methodically reach every objective step by step.

## Complete the story

In any case, every new behavior of the system should be covered by a test. So we'll need a test to include the fulfillment of the business rule, which takes us back to the acceptance level.

As this is a business rule, we'll be keeping this test afterwards.

```php
namespace App\Tests\Katas\TodoList;

use Symfony\Bundle\FrameworkBundle\Client;
use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
use Symfony\Component\HttpFoundation\Response;

class TodoListAcceptanceTest extends WebTestCase
{
    private Client $client;

    # ..

    /** @test */
    public function asUserITryToUpdateTaskWithAnEmptyTaskDescription(): void
    {
        $this->givenIHaveAddedTasks(
            [
                'Write a test that fails',
                'Write code to make the test pass',
            ]
        );

        $this->client->request(
            'PUT',
            '/api/todo/1',
            [],
            [],
            ['CONTENT-TYPE' => 'json/application'],
            json_encode(['task' => ''], JSON_THROW_ON_ERROR)
        );

        $response = $this->client->getResponse();

        self::assertEquals(400, $response->getStatusCode());

        $body = json_decode($response->getContent(), true);

        self::assertEquals('Task description should not be empty', $body['error']);
    }

    # ...
}
```


The test fails:

```
Failed asserting that 204 matches expected 400.
```

Which indicates that, right now, tasks can be created and updated with an empty description.

Now let's see how to fix this. With the available information, we don't have a clue about where to intervene.

I mean: we obviously know that we need to add a validation to the `updateDescription` that we've included in `Task`. However, skipping steps would only lead us to generating *blind spots* in the development. It's not enough to throw an exception from `Task`, we have to make sure that the appropriate captures it and reacts adequately. Proceeding systematically will help us prevent this risks.

In fact, the component that has the responsibility of communicating with the acceptance test in the first place is the controller, and as we've already seen, is the one who produces the response code that we evaluate in the acceptance test. Therefore, it's the first place where we're going to intervene. Of course, by defining a test with the expected behavior.

```php
namespace App\Tests\TodoList\Infrastructure\EntryPoint\Api;

use App\TodoList\Application\AddTaskHandler;
use App\TodoList\Application\GetTaskListHandler;
use App\TodoList\Application\MarkTaskCompletedHandler;
use App\TodoList\Application\UpdateTaskHandler;
use App\TodoList\Application\TaskListTransformer;
use App\TodoList\Infrastructure\EntryPoint\Api\TodoListController;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;
use OutOfBoundsException;
use Symfony\Component\HttpFoundation\Request;

class TodoListControllerTest extends TestCase
{
    private const TASK_DESCRIPTION = 'Task Description';
    private const COMPLETED_TASK_ID = 1;
    private const TASK_ID = 1;

    private AddTaskHandler $addTaskHandler;
    private TodoListController $todoListController;
    private GetTaskListHandler $getTaskListHandler;
    private TaskListTransformer $taskListTransformer;
    private MarkTaskCompletedHandler $markTaskCompletedHandler;
    private UpdateTaskHandler $updateTaskHandler;

    protected function setUp(): void
    {
        $this->addTaskHandler = $this->createMock(AddTaskHandler::class);
        $this->getTaskListHandler = $this->createMock(GetTaskListHandler::class);
        $this->taskListTransformer = $this->createMock(TaskListTransformer::class);
        $this->markTaskCompletedHandler = $this->createMock(MarkTaskCompletedHandler::class);
        $this->updateTaskHandler = $this->createMock(UpdateTaskHandler::class);

        $this->todoListController = new TodoListController(
            $this->addTaskHandler,
            $this->getTaskListHandler,
            $this->taskListTransformer,
            $this->markTaskCompletedHandler,
            $this->updateTaskHandler
        );
    }

    # ...
    
    /** @test */
    public function shouldFailWithBadRequestIfTaskDescriptionIsEmptyWhenUpdating(): void
    {
        $exceptionMessage = 'Task description should not be empty';
        $exception = new InvalidArgumentException($exceptionMessage);

        $this->updateTaskHandler
            ->method('execute')
            ->willThrowException($exception)
            ->with(1, '');


        $request = new Request(
            [],
            [],
            [],
            [],
            [],
            ['CONTENT-TYPE' => 'json/application'],
            json_encode(['task' => ''], JSON_THROW_ON_ERROR)
        );

        $response = $this->todoListController->modifyTask(1, $request);

        self::assertEquals(400, $response->getStatusCode());

        $body = json_decode($response->getContent(), true);

        self::assertEquals($exceptionMessage, $body['error']);
    }

}
```

When we run the test at this level, we see it fail because the exception is thrown but not controlled. We implement the exception handling just like the creation action.

```php
namespace App\TodoList\Infrastructure\EntryPoint\Api;

use App\TodoList\Application\AddTaskHandler;
use App\TodoList\Application\GetTaskListHandler;
use App\TodoList\Application\MarkTaskCompletedHandler;
use App\TodoList\Application\UpdateTaskHandler;
use App\TodoList\Application\TaskListTransformer;
use InvalidArgumentException;
use OutOfBoundsException;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use function is_string;

class TodoListController
{
    private AddTaskHandler $addTaskHandler;
    private GetTaskListHandler $getTaskListHandler;
    private TaskListTransformer $taskListTransformer;
    private MarkTaskCompletedHandler $markTaskCompletedHandler;
    private UpdateTaskHandler $updateTaskHandler;

    public function __construct(
        AddTaskHandler $addTaskHandler,
        GetTaskListHandler $getTaskListHandler,
        TaskListTransformer $taskListTransformer,
        MarkTaskCompletedHandler $markTaskCompletedHandler,
        UpdateTaskHandler $updateTaskHandler
    ) {
        $this->addTaskHandler = $addTaskHandler;
        $this->getTaskListHandler = $getTaskListHandler;
        $this->taskListTransformer = $taskListTransformer;
        $this->markTaskCompletedHandler = $markTaskCompletedHandler;
        $this->updateTaskHandler = $updateTaskHandler;
    }

    #Â ...
    
    public function markTaskCompleted(int $taskId, Request $request): Response
    {
        $payload = $this->obtainPayload($request);

        try {
            $this->markTaskCompletedHandler->execute($taskId, $payload['completed']);
        } catch (OutOfBoundsException $taskNotFound) {
            return new JsonResponse(['error' => $taskNotFound->getMessage()], Response::HTTP_NOT_FOUND);
        }

        return new JsonResponse('', Response::HTTP_OK);
    }

    public function modifyTask(int $taskId, Request $request): Response
    {
        $payload = $this->obtainPayload($request);

        try {
            $this->updateTaskHandler->execute($taskId, $payload['task']);
        } catch (\InvalidArgumentException $invalidTaskDescription) {
            return new JsonResponse(['error' => $invalidTaskDescription->getMessage()], Response::HTTP_BAD_REQUEST);
        }

        return new JsonResponse('', Response::HTTP_NO_CONTENT);
    }

    # ...
}
```

This passes the controller test. If we check the acceptance test, we see that it keeps returning the same error.

The next level is the use case, which as we've seen previously, is irrelevant because it will simply allow the exception to rise. As we know, it's `Task` who must take responsibility, so now's the time to tackle that change, defining the desired behavior in the test:

```php
namespace App\Tests\TodoList\Domain;

use App\TodoList\Domain\Task;
use PHPUnit\Framework\TestCase;
use InvalidArgumentException;

class TaskTest extends TestCase
{
    /** @test */
    public function shouldNotAllowEmptyDescription(): void
    {
        $this->expectException(InvalidArgumentException::class);

        new Task(1, '');
    }

    /** @test */
    public function shouldProvideRepresentation(): void
    {
        $expected = '[ ] 1. Task Description';
        $task = new Task(1, 'Task Description');

        $representation = $task->representedAs('[:check] :id. :description');

        self::assertEquals($expected, $representation);
    }

    /** @test */
    public function shouldMarkTaskCompleted(): void
    {
        $expected = '[â] 1. Task Description';
        $task = new Task(1, 'Task Description');
        $task->markCompleted();

        $representation = $task->representedAs('[:check] :id. :description');

        self::assertEquals($expected, $representation);
    }

    /** @test */
    public function shouldUpdateDescription(): void
    {
        $expected = '[ ] 1. New Task Description';
        $task = new Task(1, 'Task Description');
        $task->updateDescription('New Task Description');

        $representation = $task->representedAs('[:check] :id. :description');

        self::assertEquals($expected, $representation);
    }

    /** @test */
    public function shouldFailUpdatingWithInvalidDescription(): void
    {
        $this->expectException(InvalidArgumentException::class);

        $task = new Task(1, 'Task Description');
        $task->updateDescription('');
    }
}
```

As there's nothing implement, the test will fail.

We begin with a pretty obvious implementation:

```php
namespace App\TodoList\Domain;

use InvalidArgumentException;

class Task
{
    private int $id;
    private string $description;
    private bool $completed;

    public function __construct(int $id, string $description)
    {
        $this->id = $id;

        if ($description === '') {
            $exceptionMessage = 'Task description should not be empty';
            throw new InvalidArgumentException($exceptionMessage);
        }

        $this->description = $description;
        $this->completed = false;
    }

    public function id(): int
    {
        return $this->id;
    }

    public function representedAs(string $format): string
    {
        $values = [
            ':check' => $this->completed ? 'â' : ' ',
            ':id' => $this->id,
            ':description' => $this->description
        ];
        return strtr($format, $values);

    }

    public function markCompleted(): void
    {
        $this->completed = true;
    }

    public function updateDescription(string $newTaskDescription): void
    {
        if ($newTaskDescription === '') {
            $exceptionMessage = 'Task description should not be empty';
            throw new InvalidArgumentException($exceptionMessage);
        }

        $this->description = $newTaskDescription;
    }
}
```

The `Task` unit test is already in green. Before anything else, we rerun the acceptance test to se if we've solved he problem and we haven't missed any loose ends. And indeed everything works!

Nonetheless, we could refactor our solution a bit, since we're trying to maintain the same business rule in two places at once. We should unify it. To do so, we'll use auto-encapsulation, that is, we'll create a private method with which to assign and validate the description's value. This is how `Task` looks after this change.

```php
namespace App\TodoList\Domain;

use InvalidArgumentException;

class Task
{
    private int $id;
    private string $description;
    private bool $completed;

    public function __construct(int $id, string $description)
    {
        $this->id = $id;
        
        $this->setDescription($description);

        $this->completed = false;
    }

    public function id(): int
    {
        return $this->id;
    }

    public function representedAs(string $format): string
    {
        $values = [
            ':check' => $this->completed ? 'â' : ' ',
            ':id' => $this->id,
            ':description' => $this->description
        ];
        return strtr($format, $values);

    }

    public function markCompleted(): void
    {
        $this->completed = true;
    }

    public function updateDescription(string $newTaskDescription): void
    {
        $this->setDescription($newTaskDescription);
    }

    private function setDescription(string $description): void
    {
        if ($description === '') {
            $exceptionMessage = 'Task description should not be empty';
            throw new InvalidArgumentException($exceptionMessage);
        }

        $this->description = $description;
    }
}
```

And with this, we've implemented the new user story. You've probably noticed that in every case, be them new user stories, modification of features, or correction of defects, our procedure is always the same. Define the desired behavior with a test, and add the necessary production code to make it pass.

